Sources/OfflineEntitlements/PurchasedProductsFetcher.swift

```diff
@@ -0,0 +1,122 @@
+        self.unsubscribeDetectedAt = nil // best guess, StoreKit.Transaction does not provide this info.
+        self.billingIssueDetectedAt = nil // best guess, StoreKit.Transaction does not provide this info.
+        switch transaction.ownershipType {
+        case .familyShared:
+            self.ownershipType = .familyShared
+        default:
+            self.ownershipType = .purchased
+        }
+    }
+}
+
+@available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+struct PurchasedProductsManager {
+
+    func fetchPurchasedProducts() async throws -> [PurchasedSK2Product] {
+        var purchasedProductIdentifiers: [PurchasedSK2Product] = []
+
+        try await forceSyncToEnsureAllTransactionsAreAccountedFor()
+
+        for await transaction in StoreKit.Transaction.currentEntitlements {
+            switch transaction {
+            case .unverified(let unverifiedTransaction, let verificationError):
+                Logger.appleWarning(
+                    Strings.offlineEntitlements.found_unverified_transactions_in_sk2(
+                        transaction: unverifiedTransaction, error: verificationError)
+                )
+            case let .verified(verifiedTransaction):
+                purchasedProductIdentifiers.append(PurchasedSK2Product(from: verifiedTransaction))
+            }
+        }
+
+        return purchasedProductIdentifiers
+    }
+
+    private func forceSyncToEnsureAllTransactionsAreAccountedFor() async throws {
+        try await AppStore.sync()
+    }
+}
+
+@available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+extension EntitlementInfo {
+    convenience init(from purchasedSK2Product: PurchasedSK2Product, entitlementID: String) {
+        self.init(contents: .init(identifier: entitlementID,
+                                  isActive: purchasedSK2Product.isActive,
+                                  willRenew: purchasedSK2Product.willRenew,
+                                  periodType: purchasedSK2Product.periodType,
+                                  latestPurchaseDate: purchasedSK2Product.latestPurchaseDate,
+                                  originalPurchaseDate: purchasedSK2Product.originalPurchaseDate,
+                                  expirationDate: purchasedSK2Product.expirationDate,
+                                  store: purchasedSK2Product.store,
+                                  productIdentifier: purchasedSK2Product.productIdentifier,
+                                  isSandbox: purchasedSK2Product.isSandbox,
+                                  unsubscribeDetectedAt: purchasedSK2Product.unsubscribeDetectedAt,
+                                  billingIssueDetectedAt: purchasedSK2Product.billingIssueDetectedAt,
+                                  ownershipType: purchasedSK2Product.ownershipType,
+                                  verification: purchasedSK2Product.verification),
+                  rawData: [:]) // todo: should we just have the data be the transaction's raw data?
```

Comment:
Yeah ideally this matches the same format as our normal `EntitlementInfo`s, but that's going to be hard to guarantee. Given that we don't document it, I think it would be safe to have something else, like the transaction's raw data as you said 👍🏻